pkg/ccl/revertccl/revert_tenant.go

```diff
@@ -0,0 +1,136 @@
+) error {
+	execCfg := evalCtx.Planner.ExecutorConfig().(*sql.ExecutorConfig)
+
+	// These vars are set in Txn below. This transaction checks
+	// the service state of the tenant record, moves the tenant's
+	// data state to ADD, and installs a PTS for the revert
+	// timestamp.
+	//
+	// NB: We do this using a different txn since we want to be
+	// able to commit the state change during the
+	// non-transactional RevertSpans below.
+	var (
+		originalDataState mtinfopb.TenantDataState
+		tenantID          roachpb.TenantID
+		ptsCleanup        func()
+	)
+	if err := execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		tenantRecord, err := sql.GetTenantRecordByName(ctx, execCfg.Settings, txn, tenantName)
+		if err != nil {
+			return err
+		}
+		tenantID, err = roachpb.MakeTenantID(tenantRecord.ID)
+		if err != nil {
+			return err
+		}
+
+		if tenantID.Equal(roachpb.SystemTenantID) {
+			return errors.New("cannot revert the system tenant")
+		}
+
+		if tenantRecord.ServiceMode != mtinfopb.ServiceModeNone {
+			return errors.Newf("cannot revert tenant %q (%d) in service mode %s; service mode must be %s",
+				tenantRecord.Name,
+				tenantRecord.ID,
+				tenantRecord.ServiceMode,
+				mtinfopb.ServiceModeNone,
+			)
+		}
+
+		originalDataState = tenantRecord.DataState
+
+		ptsCleanup, err = protectTenantSpanWithSession(ctx, txn, execCfg, tenantID, sessionID, revertTo)
+		if err != nil {
+			return errors.Wrap(err, "protecting revert timestamp")
+		}
+
+		// Set the data state to Add during the destructive operation.
+		tenantRecord.LastRevertTenantTimestamp = revertTo
+		tenantRecord.DataState = mtinfopb.DataStateAdd
+		return sql.UpdateTenantRecord(ctx, execCfg.Settings, txn, tenantRecord)
+	}); err != nil {
+		return err
+	}
+	defer ptsCleanup()
+
+	spanToRevert := keys.MakeTenantSpan(tenantID)
+	if err := RevertSpansFanout(ctx, execCfg.DB, evalCtx.JobExecContext.(sql.JobExecContext),
+		[]roachpb.Span{spanToRevert},
+		revertTo,
+		false, /* ignoreGCThreshold */
+		RevertDefaultBatchSize,
+		nil /* onCompletedCallback */); err != nil {
+		return err
+	}
+
+	return execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		tenantRecord, err := sql.GetTenantRecordByName(ctx, execCfg.Settings, txn, tenantName)
+		if err != nil {
+			return err
+		}
+		tenantRecord.DataState = originalDataState
+		return sql.UpdateTenantRecord(ctx, execCfg.Settings, txn, tenantRecord)
+	})
+}
+
+func protectTenantSpanWithSession(
+	ctx context.Context,
+	txn isql.Txn,
+	execCfg *sql.ExecutorConfig,
+	tenantID roachpb.TenantID,
+	sessionID clusterunique.ID,
+	timestamp hlc.Timestamp,
+) (func(), error) {
+	ptsRecordID := uuid.MakeV4()
+	ptsRecord := sessionprotectedts.MakeRecord(
+		ptsRecordID,
+		[]byte(sessionID.String()),
+		timestamp,
+		ptpb.MakeTenantsTarget([]roachpb.TenantID{tenantID}),
+	)
+	log.Infof(ctx, "protecting timestamp: %#+v", ptsRecord)
```

Comment:
Existing code, but perhaps we should clean up this log message a bit. Perhaps printing just the tenant ID and timestamp rather than the entire record struct?